board/opentrons/ot2: Work around RV-3028 problems via artificial delay

SyntaxColoring opened the pull request:
# Overview

This PR works around problems with the OT-2 Refresh's RV-3028 realtime clock (RTC).

It patches the RTC's device driver in the Linux kernel. Whenever something tries to read the RTC's current time, we add an artificial delay.

See the source comment for details about how this works and why we do it. See Jira RSS-9 for more background on the investigation.

# Review requests

It will be easier to view the code changes here:

https://github.com/SyntaxColoring/linux-rv3028-hacking/compare/opentrons-buildroot-v1.7.0-with-patches...SyntaxColoring:linux-rv3028-hacking:rv3028_hacking

Those are the Git commits that I generated this patch file from.

* [ ] This is baby's first kernel patch. Does my code look correct?
* [ ] In terms of buildroot config, is this patch file thing the best way to apply these changes?
* [x] At the I2C level, does this actually have the intended effect of inserting a 2 ms delay before each read of the time registers?
    * Verified by @SyntaxColoring with a logic analyzer.
* [x] Does this actually let the `hwclock` commands used by `ot2-factory-tools` succeed on problematic boards?
    * Verified by @SyntaxColoring and by the Shenzhen team.
    * Technically should be verified again with the `usleep_range()` changes discussed in the review comments below, but I'm pretty confident that it's fine. I'll double-check again when we release this to the factory, just to be safe.

# Risk analysis

This is pretty risky. We're messing with kernel code, so there's really no bound to the damage we can do if we screw something up. :)

# What merging this PR means

This PR will be merged into `opentrons-develop`. So, it will get pushed into the field whenever we tag our next buildroot release and whenever the next robot software release comes out and bundles it. There's no rush for this, because the problem does not affect any robots currently in the field—it only affects robots on the assembly line.

To get this workaround to robots on the assembly line, we will probably need to backport this fix to [our v1.7 buildroot release](https://github.com/Opentrons/buildroot/releases/tag/v1.7.0), since that's what the factory is using. We'll do this separately after this PR passes code review.

SyntaxColoring committed:
board/opentrons/ot2: Work around RV-3028 problems via artificial delay

sfoster1 reviewed (changes requested):
Couple things:
1. Let's not put the big comment in the patch file, for two reasons. First, kernel development practice is that explanations of this type live in commit messages or documentation rather than source comments; this keeps the code itself quicker to read while keeping the explanations around in a place that is still easy to read. Second, the way we use the kernel, we're really not keeping the source around. The thing we wrote that changes the behavior is the patch file. With these factors combined, I'd say the explanation wants to be in the message-part of the patch. Also the NOTE() thing isn't really a thing in kernel style comments.
2. Definitely do not use `udelay` for this. `udelay` is a jiffie-estimator spinloop that blocks scheduler yielding. There is a very strong likelihood that this code will hardstop the world, or at least a core, for 2ms every time something tries to read the time. As the timer-howto doc mentions, `usleep_range` is the right tool to use here given the length of the delay since it allows for scheduler interruptions. `rv3028_get_time` is used as a driver ops bind, part of the public interface for rtcs, so i'm pretty sure it's not in an atomic context - those tend to occur inside specifically locked code sections or callbacks.

All that said, a source patch like this is indeed the correct way to do this (though, we'll probably have to regenerate it when we change kernel versions; that won't be hard since that code doesn't appear to have changed significantly, like, ever), and inserting a sleep in there probably will fix the problem.

vegano1 commented:
@SyntaxColoring 
You pointed out before that timestamp0_show might also have this issue, although I'm not sure how frequently this function gets called or if its even a problem. But it might make sense to make this same change there as well.

SyntaxColoring committed:
Switch to usleep_range() and move explanation into patch description.

SyntaxColoring committed:
Explain where 2-4 ms comes from.

SyntaxColoring commented:
> @SyntaxColoring You pointed out before that timestamp0_show might also have this issue, although I'm not sure how frequently this function gets called or if its even a problem. But it might make sense to make this same change there as well.

Yeah, after thinking about that some more, I'm deliberately leaving it alone, for now.

To give others some background, I was worried about this other line in the driver file:

https://github.com/SyntaxColoring/linux-rv3028-hacking/blob/1fa2f95daf21e3d4d4b2866ad7829abc8c1a3af7/drivers/rtc/rtc-rv3028.c#L121-L122

Which looks like an identical `regmap_bulk_read()` call to the one that we're inserting a sleep before, so maybe we need a second sleep there. But it's actually different: notice the register isn't `RV3028_SEC`, but rather `RV3028_TS_SEC`. (`TS` refers to the chip's timestamping function, which appears totally separate from its time*keeping* function.)

I'm leaving it alone because:

* Other than my newfound distrust of these chips in general, I don't have any reason to believe that `timestamp0_show()` is broken.
* As far as I know, nothing in our system uses `timestamp0_show()`. I wouldn't even know how to test it.

SyntaxColoring committed:
Fix whitespace problems introduced in copy-paste?

SyntaxColoring committed:
Reformat patch.

- Replace spaces with tabs.
  The spaces were introduced erroneously by my editor
  and were preventing the patch from applying cleanly.
- Include default Git junk at the top, for consistency.

sfoster1 reviewed (approved):
Great! Looks good to me. Thanks for making the changes!
